panic: runtime error: invalid memory address or nil pointer deference (kiln-prysm-besu)

# 🐞 Bug Report

### Description

prysm-besu panics and exits on startup.  This has happened twice that I've noticed now, and restarting prysm beacon allows it to resume syncing.

### Has this worked before in a previous version?

This only started happening recently.

## 🔬 Minimal Reproduction

`bazelisk run //beacon-chain -- --genesis-state $PWD/../genesis.ssz --datadir $HOME/eth/kiln-prysm-besu --http-web3provider=http://localhost:8578 --min-sync-peers=1 --chain-config-file $PWD/../config.yaml --bootstrap-node "enr:-Iq4QMCTfIMXnow27baRUb35Q8iiFHSIDBJh6hQM5Axohhf4b6Kr_cOCu0htQ5WvVqKvFgY28893DHAg8gnBAXsAVqmGAX53x8JggmlkgnY0gmlwhLKAlv6Jc2VjcDI1NmsxoQK6S-Cii_KmfFdUJL2TANL3ksaKUnNXvTCv1tLwXs0QgIN1ZHCCIyk" --p2p-tcp-port 13468 --p2p-udp-port 13468 --p2p-host-ip "1.2.3.4" --rpc-port 4002 --grpc-gateway-port 3502 --monitoring-port 8084 --suggested-fee-recipient "0xA721da3d06813E764E94A23E377cC3E1729fCFD5" --jwt-secret=/tmp/jwtsecret --dev`

`bazelisk run //validator -- --chain-config-file ~/Downloads/merge-testnets/kiln/config.yaml --wallet-password-file ~/eth/kiln-prysm-besu/smashword --wallet-dir ~/eth/kiln-prysm-besu/prysm-wallet-v2 --graffiti "prysm-besu seamonkey.tech" --beacon-rpc-gateway-provider 127.0.0.1:3502 --beacon-rpc-provider 127.0.0.1:4002 --monitoring-port 8083 --suggested-fee-recipient "0xA721da3d06813E764E94A23E377cC3E1729fCFD5"`


## 🔥 Error

<pre><code>[2022-04-18 21:24:02]  INFO initial-sync: Processing block batch of size 30 starting from  0xd23e6cd9... 276929/276995 - estimated time remaining 44s blocksPerSecond=1.5 peers=5
[2022-04-18 21:24:12]  WARN initial-sync: Batch is not processed error=payload status is INVALID
github.com/prysmaticlabs/prysm/beacon-chain/powchain.init
        beacon-chain/powchain/errors.go:30
runtime.doInit
        GOROOT/src/runtime/proc.go:6498
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.main
        GOROOT/src/runtime/proc.go:238
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
could not notify forkchoice update from execution engine
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).notifyForkchoiceUpdate
        beacon-chain/blockchain/execution_engine.go:94
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).onBlockBatch
        beacon-chain/blockchain/process_block.go:438
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveBlockBatch
        beacon-chain/blockchain/receive_block.go:82
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processBatchedBlocks
        beacon-chain/sync/initial-sync/round_robin.go:283
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processFetchedData
        beacon-chain/sync/initial-sync/round_robin.go:133
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).syncToFinalizedEpoch
        beacon-chain/sync/initial-sync/round_robin.go:86
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).roundRobinSync
        beacon-chain/sync/initial-sync/round_robin.go:49
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).Start
        beacon-chain/sync/initial-sync/service.go:104
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
could not process block in batch
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveBlockBatch
        beacon-chain/blockchain/receive_block.go:84
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processBatchedBlocks
        beacon-chain/sync/initial-sync/round_robin.go:283
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processFetchedData
        beacon-chain/sync/initial-sync/round_robin.go:133
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).syncToFinalizedEpoch
        beacon-chain/sync/initial-sync/round_robin.go:86
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).roundRobinSync
        beacon-chain/sync/initial-sync/round_robin.go:49
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).Start
        beacon-chain/sync/initial-sync/service.go:104
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
[2022-04-18 21:24:12]  INFO initial-sync: Processing block batch of size 33 starting from  0xfd742d0d... 276960/276996 - estimated time remaining 11s blocksPerSecond=3.1 peers=6
[2022-04-18 21:24:13]  WARN initial-sync: Batch is not processed error=payload status is INVALID
github.com/prysmaticlabs/prysm/beacon-chain/powchain.init
        beacon-chain/powchain/errors.go:30
runtime.doInit
        GOROOT/src/runtime/proc.go:6498
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.doInit
        GOROOT/src/runtime/proc.go:6475
runtime.main
        GOROOT/src/runtime/proc.go:238
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
could not notify forkchoice update from execution engine
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).notifyForkchoiceUpdate
        beacon-chain/blockchain/execution_engine.go:94
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).onBlockBatch
        beacon-chain/blockchain/process_block.go:438
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveBlockBatch
        beacon-chain/blockchain/receive_block.go:82
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processBatchedBlocks
        beacon-chain/sync/initial-sync/round_robin.go:283
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processFetchedData
        beacon-chain/sync/initial-sync/round_robin.go:133
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).syncToFinalizedEpoch
        beacon-chain/sync/initial-sync/round_robin.go:86
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).roundRobinSync
        beacon-chain/sync/initial-sync/round_robin.go:49
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).Start
        beacon-chain/sync/initial-sync/service.go:104
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
could not process block in batch
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveBlockBatch
        beacon-chain/blockchain/receive_block.go:84
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processBatchedBlocks
        beacon-chain/sync/initial-sync/round_robin.go:283
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processFetchedData
        beacon-chain/sync/initial-sync/round_robin.go:133
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).syncToFinalizedEpoch
        beacon-chain/sync/initial-sync/round_robin.go:86
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).roundRobinSync
        beacon-chain/sync/initial-sync/round_robin.go:49
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).Start
        beacon-chain/sync/initial-sync/service.go:104
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1581
[2022-04-18 21:24:13]  INFO initial-sync: Synced to finalized epoch - now syncing blocks up to current head currentSlot=276996 syncedSlot=276928
[2022-04-18 21:24:15]  WARN blockchain: Could not update head error=head at slot 276960 with weight 0 is not eligible, finalizedEpoch 8653 != 8654, justifiedEpoch 8654 != 8655
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x136bc81]

goroutine 313 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).onBlock(0xc000542460, {0x1edae38, 0xc005334120}, {0x1f0f3d0, 0xc004f61f40}, {0xd2, 0x3e, 0x6c, 0xd9, 0x7b, ...})
        beacon-chain/blockchain/process_block.go:222 +0xb41
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveBlock(0xc000542460, {0x1edad90, 0xc004f5fec0}, {0x1f0f3d0, 0xc004e240a0}, {0xd2, 0x3e, 0x6c, 0xd9, 0x7b, ...})
        beacon-chain/blockchain/receive_block.go:39 +0x15f
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processBlock(0xc00befefc0, {0x1edad90, 0xc004f5fec0}, {0x5, 0x73, 0x3031ec0}, {0x1f0f3d0, 0xc004e240a0}, 0xc00ea20f68)
        beacon-chain/sync/initial-sync/round_robin.go:239 +0x39d
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).processFetchedDataRegSync(0xc00befefc0, {0x1edad90, 0xc004f5fec0}, {0x0, 0x0, 0x3031ec0}, 0x439e0, 0xc0053219b0)
        beacon-chain/sync/initial-sync/round_robin.go:146 +0x325
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).syncToNonFinalizedEpoch(0xc00befefc0, {0x1edad90, 0xc004f5fec0}, {0x16d99a0, 0x2e0e940, 0x3031ec0})
        beacon-chain/sync/initial-sync/round_robin.go:114 +0x1d9
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).roundRobinSync(0xc00befefc0, {0xe00000004, 0xc003e61a48, 0x3031ec0})
        beacon-chain/sync/initial-sync/round_robin.go:60 +0x239
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*Service).Start(0xc00befefc0)
        beacon-chain/sync/initial-sync/service.go:104 +0x8f5
created by github.com/prysmaticlabs/prysm/runtime.(*ServiceRegistry).StartAll
        runtime/service_registry.go:46 +0x225
</code></pre>


## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Linux Mint 20.3
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
d2f4a8cc7c53a684ab2f420f5ed6dc527143a477
  </code>
</pre>

--- a/beacon-chain/blockchain/process_block.go
+++ b/beacon-chain/blockchain/process_block.go
@@ -207,20 +207,7 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.SignedBeaconBlo
 	if err != nil {
 		log.WithError(err).Warn("Could not update head")
 	}
-	headBlock, err := s.cfg.BeaconDB.Block(ctx, headRoot)
-	if err != nil {
-		return err
-	}
-	headState, err := s.cfg.StateGen.StateByRoot(ctx, headRoot)
-	if err != nil {
-		return err
-	}
-	if _, err := s.notifyForkchoiceUpdate(ctx, headState, headBlock.Block(), headRoot, bytesutil.ToBytes32(finalized.Root)); err != nil {
-		return err
-	}
-	if err := s.saveHead(ctx, headRoot, headBlock, headState); err != nil {
-		return errors.Wrap(err, "could not save head")
-	}
+	s.notifyEngineIfChangedHead(ctx, headRoot)
 
 	if err := s.pruneCanonicalAttsFromPool(ctx, blockRoot, signed); err != nil {
 		return err

